### DISCUSSION TITLE
GPU is idle when enabled

### DISCUSSION THREAD

**Discussion by Sebekerga** [milvus-io/milvus#27803] [Q&A and General discussion] [answered]
I've setup Milvus according to [this guide](https://milvus.io/docs/install_standalone-gpu-docker.md), no errors, everything works, Milvus even is shown in `nvtop` to have some memory allocated. But whenever I run query from python, GPU memory stays the same, but my CPU is being hammered. 

I setup my index like so:
```python
index = {
    "index_type": "GPU_IVF_FLAT",
    "metric_type": "COSINE",
    "params": { "nlist": 2048 },
}
```

**Comment by yhmo** [comment, marked as answer]
- If the row count is less than 1024, milvus will use brute-force search instead of index search. How many rows in the collection? 
- If the index is not ready(index node work is not finished), milvus will use brute-force search.
- GPU index doesn't support cosine metric, only support L2 and IP metric.

Try this index:
```
index = {
    "index_type": "GPU_IVF_FLAT",
    "metric_type": "L2",
    "params": { "nlist": 2048 },
}
```

Insert more rows into the collection, millions of rows. Call create_index() after you insert the millions of rows. And try search() again.

**Reply by Sebekerga** [reply]
Thanks for reply. 
I have ~150k rows uploaded and I query 3k vectors at a time.

> If the index is not ready(index node work is not finished), milvus will use brute-force search.

How can I wait for it to load with `pymilvus` client? Also, I create index and then insert rows, but GPU memory stays constant during inserting at around 6GB. 

> GPU index doesn't support cosine metric, only support L2 and IP metric.

Let me try with L2, thanks

**Reply by Sebekerga** [reply]
I tried with L2, nothing changed, still full load on CPU. I've also tried on another machine, and the GPU works, even with COSINE metric (assuming it might just fall back to L2 or smth, but regardless), same python code.

**Reply by Sebekerga** [reply]
I've tried indexing after loading rows, given L2 as a metric, it works, thanks for help @yhmo
